### PR TITLE
[persist/storage] Spawn a task for the finalize-shard operation

### PR DESCRIPTION
**Pull request by bkirwi**
### Motivation

Known improvement/bug: #22508

### Tips for reviewer

The core annoyance here is that we need to reference the stash collection if we want to actually clean stuff up, but the task can't hang on to self. I tried a few ways to get around this, but the simplest is just to fork off a task for the actual upper advancement, and rely on some future invocation of the method to clean out the stash.

My biggest worry here is forking off too many tasks at once and chewing resources. If that worries you too I will add a semaphore.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/persist-client/src/lib.rs [hunk 1]
```diff
@@ -388,7 +388,7 @@ impl PersistClient {
         ))
     }
 
-    /// [Self::open], but returning only a [/eadHandle].
```
Comment (danhhz) [inline src/persist-client/src/lib.rs:391]: Wild, I'm surprised this didn't fail our docs CI

**Review by guswynn — COMMENTED** [review]
the use of the task without a semaphore seems fine to me!
however I don't think I understand persist semantics enough to comment on how the state machine works here!

File: src/storage-controller/src/lib.rs [hunk 2]
```diff
@@ -2657,21 +2657,41 @@ where
...
+                                    .await;
+
                             // Rather than error, just leave this shard as one to finalize later.
-                            .ok()?
-                            .ok()?;
+                            match result {
+                                Err(usage) => {
+                                    error!("invalid usage while finalizing shard: {usage:?}")
+                                }
+                                Ok(Err(mismatch)) => {
```
Comment (danhhz) [inline src/storage-controller/src/lib.rs:2687]: This call is using `append`, which automatically retries if the upper advances. Looks like `append` returns this error if you called it with an upper that is too far ahead of the actual one (e.g. the real upper is 5, but your append call is `(7, ())`. We're using `write_handle.upper().clone()` above, which means this should never happen and we could probably be more aggressive here if we liked (i.e. error or panic)

**Review by danhhz — APPROVED** [review]
I like this approach to the `&mut self`!

Something I just thought of, doesn't have to be for this PR, is we might want to put a really long timeout on the compare_and_append call here so we don't get perma-stuck in the "crdb query is too big" loop?

File: src/storage-controller/src/lib.rs (hunk 2 above)
Comment (bkirwi) [reply src/storage-controller/src/lib.rs]: Point! I bumped to `warn` for now, but I'd be happy to make it even more aggressive in the future.

**Comment by bkirwi** [comment]
I added a fifteen-minute timeout to the call, which feels like it should be sufficiently really long... let me know if I should bump it further!

**Comment by danhhz** [comment]
SGTM!

**Pull request merged**